lib/api/v3/utilities/custom_field_injector.rb

```diff
@@ -282,20 +282,34 @@ def property_value_setter_for(custom_field)
           }
         end
 
+        # rubocop:disable Metrics/AbcSize
         def allowed_users_href_callback
+          # Careful to not alter the static_filters object here.
+          # It is made available in the closure (which is class level) and would thus
+          # keep the appended filters between requests.
           static_filters = allowed_users_static_filters
           instance_filters = method(:allowed_users_instance_filter)
 
-          ->(*) {
-            # Careful to not alter the static_filters object here.
-            # It is made available in the closure (which is class level) and would thus
-            # keep the appended filters between requests.
-            filters = static_filters + instance_filters.call(represented)
+          represented_is_existent_instance = lambda do |represented, clazz|
+            represented.respond_to?(:model) &&
+              represented.model.is_a?(clazz) &&
+              represented.model.id.present?
+          end
 
-            api_v3_paths.path_for(:principals, filters:, page_size: -1)
+          ->(*) {
+            if represented_is_existent_instance.(represented, Project)
+              api_v3_paths.available_assignees_in_project(represented.id)
+            elsif represented_is_existent_instance.(represented, WorkPackage)
+              api_v3_paths.available_assignees_in_work_package(represented.id)
```

Comment:
This change reduces the possible values to those that have the `work_package_assigned` permission. This includes those that have the work package shared with them but excludes those that are members of the project but have a different set of permissions. Before, all project members where possible values. On top of this, only if the work package is shared with the edit permission (which includes `work_package_assigned`), will the shared with user show up in the list.

The intended outcome would be to allow all project members and on top of those also allow users which have a work package membership. 

To that end, a specific end point would probably be the best solution. 

What is unsettling about the code change is that it indicates a lack of validations when actually creating a custom value for the custom field. There doesn't appear to be any constraint. This weakness existed before the PR.

Reply:
Ahoi @ulferts ,

should we take that discussion into a synchronous call? I'd argue, that you shouldn't be able to assign users to a work package custom field of type user, if this user does not have the `work_package_assigned` permission.

So, this is in the end a product decision: Is the assignee set the same for the default `Assignee` field and any custom field of type user (easy solution)? Or are there differences and if yes, what kind?

Reply:
> What is unsettling about the code change is that it indicates a lack of validations when actually creating a custom value for the custom field. There doesn't appear to be any constraint. This weakness existed before the PR.

How does it indicate that? The change only changes the `available_values` in the schema response of the APIv3. It does not change the actual update process/validation. Is it true, that there is no validation? Could you assign any user?